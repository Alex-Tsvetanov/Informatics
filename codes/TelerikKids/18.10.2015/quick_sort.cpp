#include <iostream>
#include <random>

using namespace std;

template <typename T>
T* get_pivot (T* a, T* b)
{
	T* masterPivot = a;
	T averrage = 0;
	for (T* it = a ; it != b ; it ++)
		averrage += (*it);
	averrage /= (b - a);
	for (T* it = a ; it != b ; it ++)
		if (abs ((*masterPivot) - averrage) > abs ((*it) - averrage))
			masterPivot = it;

	return masterPivot;
}

template <typename T>
T* fillArr (T* newArr, T* pivot, T* s, T* e)
{
	T* indOfPivot;
	size_t ind = 0;
	for (T* it = s ; it != e ; it ++)
		if ((*it) < (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	indOfPivot = newArr + ind;
	for (T* it = s ; it != e ; it ++)
		if ((*it) >= (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	return indOfPivot;
}

template <typename T>
void copyArr (T* newArr, T* s, T* e)
{
	T* newArrIt = newArr;
	for (T* it = s ; it != e ; it ++, newArrIt ++)
		*it = *newArrIt;
}

template <typename T>
void quick_sort (T* s, T* e)
{
	if (s + 1 == e)
		return ;
	if (s == e)
		return ;
	T* newArr = new T [e - s];
	T* pivot = fillArr (newArr, get_pivot (s, e), s, e);
	//for (T* it = s ; it != e ; it ++)
	//	cout << *it << " ";
	//cout << "\n" << *pivot << "\n----------------------------" << endl;
	quick_sort (newArr, pivot + 1);
	quick_sort (pivot + 1, newArr + (e - s));
	copyArr (newArr, s, e);
	delete[] newArr;
}

int main ()
{
	size_t SIZE;
	cin >> SIZE;
	int* arr = new int [SIZE];
	for (size_t i = 0 ; i < SIZE ; i ++)
		cin >> arr [i];
	quick_sort (arr + 0, arr + SIZE);
	for (size_t i = 0 ; i < SIZE ; i ++)
		cout << arr [i] << " ";
	cout << "\n";
}
