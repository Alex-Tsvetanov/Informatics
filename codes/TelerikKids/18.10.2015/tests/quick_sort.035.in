78
27 42 59 52 21 67 52 91 69 75 98 15 55 15 67 46 85 47 36 66 16 89 42 64 28 91 2 42 82 15 81 0 37 52 36 82 33 17 37 3 81 63 60 100 67 28 56 73 80 13 68 37 5 70 46 41 94 30 64 21 57 26 95 16 52 1 14 90 29 70 34 28 58 8 48 51 1 85 
