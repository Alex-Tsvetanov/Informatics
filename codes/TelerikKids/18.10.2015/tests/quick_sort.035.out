0 1 1 2 3 5 8 13 14 15 15 15 16 16 17 21 21 26 27 28 28 28 29 30 33 34 36 36 37 37 37 41 42 42 42 46 46 47 48 51 52 52 52 52 55 56 57 58 59 60 63 64 64 66 67 67 67 68 69 70 70 73 75 80 81 81 82 82 85 85 89 90 91 91 94 95 98 100 
